File: INCLService/SchichtUtilLib.pas
unit SchichtUtilLib;

interface

uses Classes, SysUtils;

type
  TStillstandEintrag = class
  public
    Nr : integer;
    Kommt: TDateTime;
    Geht: TDateTime;
    GrundNr: Integer;
    Geplant: Boolean;
    Maschnr : Integer;
    Gruppe: Integer; // 0 -> Anlagenausfall, 1 -> Rüsten 2 ->  Logistik 3 -> ungebucht
    Stillstand : string;
    function CopyMe : TStillstandEintrag;
  end;

  TStillstandEintragsListe = class(TList)
  private
    function getItem(index: Integer): TStillstandEintrag;
    procedure setItem(index: Integer; const Value: TStillstandEintrag);

  public
    property Items[index: Integer]: TStillstandEintrag read getItem write setItem;

    function getMaschNrsString : string;
    function GetByMaschNr(aMaschNr : Integer) : TStillstandEintragsListe;
    function Add(aStillstandEintrag: TStillstandEintrag): Integer;
    function AddRaw(aStillstandEintrag: TStillstandEintrag): Integer;
    function GetDauerByMaschNr(aMaschNr : Integer) : Integer;
    function GetDauerByMaschNrFromDate(aFromDate : TDatetime; aMaschNr : Integer) : Integer;
    function GetOpenByMaschNr(aMaschNr : Integer) : TStillstandEintrag;
    procedure Clear;
    destructor Destroy;

  end;

  TStartStopEintrag = class
  private
    function CopyMe : TStartStopEintrag;
  public
    AuftragNr: string;
    RuestStart: TDateTime;
    Start: TDateTime;
    Stop: TDateTime;

  end;

  TStartStopEintragsListe = class(TList)
  private
    function getItem(index: Integer): TStartStopEintrag;
    procedure setItem(index: Integer; const Value: TStartStopEintrag);

  public
    property Items[index: Integer]: TStartStopEintrag read getItem write setItem;
    function Add(aStartStopEintrag: TStartStopEintrag): Integer;
    function GetByBetriebsauftragNr(aBANr : string) : TStartStopEintragsListe;
    procedure Clear;
    destructor Destroy;
  end;
type
  TSignalLogEintrag = class
  public
    maschnr: integer;
    wert : integer;
    Start: TDateTime;
    Stop: TDateTime;

    function CopyMe : TSignalLogEintrag;
  end;
type
  TSignalLogEintragListe = class(TList)
  private
    function getItem(index: Integer): TSignalLogEintrag;
    procedure setItem(index: Integer; const Value: TSignalLogEintrag);
  public
    property Items[index: Integer]: TSignalLogEintrag read getItem write setItem;
    function Add(aSignalLogEintrag: TSignalLogEintrag): Integer;
    function GetByMaschNr(aMaschNr : Integer) : TSignalLogEintragListe;
    procedure Clear;
    destructor Destroy;
  end;

implementation

{ TStillstandEintragsListe }

function TStillstandEintragsListe.Add(
  aStillstandEintrag: TStillstandEintrag): Integer;
var found : boolean;
    i : integer;
begin
  for i := 0 to Self.Count-1 do
  begin
    found := Items[i].Nr = aStillstandEintrag.Nr;
    if found then
    begin
      Result := i;
      exit;
    end;
  end;
  // Um zu vermeiden dass doppelte drin sein könnten wird geprüft
  Result := TList(Self).Add(aStillstandEintrag);
end;

function TStillstandEintragsListe.AddRaw(aStillstandEintrag: TStillstandEintrag): Integer;
begin
  Result := TList(Self).Add(aStillstandEintrag);
end;


function TStillstandEintragsListe.GetDauerByMaschNr(aMaschNr : Integer) : Integer;
var dauer : Extended;
  i : Integer;
    stillende : extended;
begin
  dauer := 0;
  for i := 0 to Count-1 do
  begin
    if Items[i].Maschnr = aMaschNr then
    begin
      stillende := Items[i].Geht;
      if Items[i].geht < 1 then
        stillende := Now;
      if Items[i].kommt < stillende then
        dauer := dauer + ((stillende - Items[i].kommt)*1440);
    end;
  end;
  result := round(dauer);
end;

function TStillstandEintragsListe.GetDauerByMaschNrFromDate(aFromDate : TDatetime; aMaschNr : Integer) : Integer;
var dauer : Extended;
  i : Integer;
    stillgeht : extended;
    stillkommt : extended;
begin
  dauer := 0;
  for i := 0 to Count-1 do
  begin
    if Items[i].Maschnr = aMaschNr then
    begin
      stillgeht := Items[i].Geht;
      stillkommt := items[i].Kommt;
      if stillgeht < 1 then
        stillgeht := Now;
      if stillkommt < aFromDate then
        stillkommt := aFromDate;
      if stillkommt > stillgeht then
        stillkommt := stillgeht;
      dauer := dauer + ((stillgeht - stillkommt)*1440);
    end;
  end;
  result := round(dauer);
end;


procedure TStillstandEintragsListe.Clear;
begin
  while Self.Count > 0 do
  begin
    Self.Items[0].Free;
    Self.Delete(0);
  end;
  inherited;
end;

destructor TStillstandEintragsListe.Destroy;
begin
  while Self.Count > 0 do
  begin
    Self.Items[0].Free;
    Self.Delete(0);
  end;
  inherited;
end;

function TStillstandEintragsListe.GetByMaschNr(
  aMaschNr: Integer): TStillstandEintragsListe;
var i : Integer;
begin
  result := TStillstandEintragsListe.Create;
  for i := 0 to self.Count-1 do
  begin
    if self.Items[i].maschnr = aMaschnr then
    begin
//      result.Add(self.Items[i].CopyMe);
        result.Add(self.Items[i]);
    end;
  end;

end;


function TStillstandEintragsListe.getItem(
  index: Integer): TStillstandEintrag;
begin
  Result := TStillstandEintrag(TList(Self).Items[index]);

end;

function TStillstandEintragsListe.getMaschNrsString: string;
var s : string;
    i : Integer;
begin
  s := '';
  for i := 0 to self.Count-1 do
  begin
    if Pos(' ' + IntTostr(self.Items[i].Maschnr) + ',', s) = 0 then
      s := s + ' ' + IntTostr(self.Items[i].Maschnr) + ',';
  end;
  if s <> '' then
    system.Delete(s, length(s),1);
  result := s;
end;

procedure TStillstandEintragsListe.setItem(index: Integer;
  const Value: TStillstandEintrag);
begin
  Self[index] := Value;
end;

function TStillstandEintragsListe.GetOpenByMaschNr(
  aMaschNr: Integer): TStillstandEintrag;
var i : Integer;
begin
  result := nil;
  for i := 0 to self.Count-1 do
  begin
    if (self.Items[i].maschnr = aMaschnr)and (self.Items[i].Geht < 1) then
    begin
      result :=self.Items[i];
      exit;
    end;
  end;
end;

{ TStartStopEintragsListe }

function TStartStopEintragsListe.Add(
  aStartStopEintrag: TStartStopEintrag): Integer;
begin
  Result := TList(Self).Add(aStartStopEintrag);
end;

procedure TStartStopEintragsListe.Clear;
begin
  while Self.Count > 0 do
  begin
    Self.Items[0].Destroy;
    Self.Delete(0);
  end;
  inherited;

end;

destructor TStartStopEintragsListe.Destroy;
begin
  while Self.Count > 0 do
  begin
    Self.Items[0].Destroy;
    Self.Delete(0);
  end;
  inherited;
end;

function TStartStopEintragsListe.GetByBetriebsauftragNr(
  aBANr: string): TStartStopEintragsListe;
var i : Integer;
begin
  result := TStartStopEintragsListe.Create;
  for i := 0 to self.Count-1 do
  begin
    if self.Items[i].AuftragNr = aBANr then
    begin
//      result.Add(self.Items[i].CopyMe);
      result.Add(self.Items[i]);
    end;
  end;
end;

function TStartStopEintragsListe.getItem(
  index: Integer): TStartStopEintrag;
begin
  Result := TStartStopEintrag(TList(Self).Items[index]);

end;

procedure TStartStopEintragsListe.setItem(index: Integer;
  const Value: TStartStopEintrag);
begin
  Self[index] := Value;
end;

{ TSignalLogEintragListe }

function TSignalLogEintragListe.Add(
  aSignalLogEintrag: TSignalLogEintrag): Integer;
begin
  Result := TList(Self).Add(aSignalLogEintrag);

end;

procedure TSignalLogEintragListe.Clear;
begin
  while Self.Count > 0 do
  begin
    Self.Items[0].Destroy;
    Self.Delete(0);
  end;
  inherited;

end;

destructor TSignalLogEintragListe.Destroy;
begin
  while Self.Count > 0 do
  begin
//    Self.Items[0].Destroy;
    Self.Items[0].Free;
    Self.Delete(0);
  end;
  inherited;

end;

function TSignalLogEintragListe.GetByMaschNr(
  aMaschNr: Integer): TSignalLogEintragListe;
var i : Integer;
begin
  result := nil;
        result := TSignalLogEintragListe.Create;
  for i := 0 to self.Count-1 do
  begin
    if self.Items[i].maschnr = aMaschnr then
    begin
//      if result = nil then
  //      result := TSignalLogEintragListe.Create;
//      result.Add(self.Items[i].CopyMe);
      result.Add(self.Items[i]);
    end;
  end;
end;

function TSignalLogEintragListe.getItem(index: Integer): TSignalLogEintrag;
begin
  Result := TSignalLogEintrag(TList(Self).Items[index]);

end;

procedure TSignalLogEintragListe.setItem(index: Integer;
  const Value: TSignalLogEintrag);
begin
  Self[index] := Value;

end;

{ TSignalLogEintrag }

function TSignalLogEintrag.CopyMe: TSignalLogEintrag;
begin
  Result := TSignalLogEintrag.Create;
  Result.maschnr := self.maschnr;
  Result.wert := self.wert;
  Result.Start := self.Start;
  Result.Stop := self.Stop;
end;

{ TStillstandEintrag }

function TStillstandEintrag.CopyMe: TStillstandEintrag;
begin
  result := TStillstandEintrag.Create;
  result.Kommt := self.Kommt;
  result.Geht := self.Geht;
  result.GrundNr := self.GrundNr;
  result.Geplant := self.Geplant;
  result.Maschnr := self.Maschnr;
  result.Gruppe := self.Gruppe;
  result.Stillstand := self.Stillstand;
end;

{ TStartStopEintrag }

function TStartStopEintrag.CopyMe: TStartStopEintrag;
begin
  result := TStartStopEintrag.Create;
  result.Start := self.Start;
  result.Stop := self.Stop;
  result.AuftragNr := self.AuftragNr;
  result.RuestStart := self.RuestStart;
end;

end.

File: INCLService_CSharp/SchichtUtilLib.cs
using System;
using System.Collections.Generic;

namespace INCLService_CSharp
{
    public class TStillstandEintrag
    {
        public int Nr { get; set; } = 0;
        public DateTime Kommt { get; set; } = DateTime.MinValue;
        public DateTime Geht { get; set; } = DateTime.MinValue;
        public int GrundNr { get; set; } = 0;
        public bool Geplant { get; set; } = false;
        public int Maschnr { get; set; } = 0;
        public int Gruppe { get; set; } = 0; // 0 -> Anlagenausfall, 1 -> Rüsten 2 -> Logistik 3 -> ungebucht
        public string Stillstand { get; set; } = "";

        public TStillstandEintrag CopyMe()
        {
            return new TStillstandEintrag
            {
                Nr = this.Nr,
                Kommt = this.Kommt,
                Geht = this.Geht,
                GrundNr = this.GrundNr,
                Geplant = this.Geplant,
                Maschnr = this.Maschnr,
                Gruppe = this.Gruppe,
                Stillstand = this.Stillstand
            };
        }
    }

    public class TStillstandEintragsListe : List<TStillstandEintrag>
    {
        public new TStillstandEintrag this[int index]
        {
            get => base[index];
            set => base[index] = value;
        }

        public int Add(TStillstandEintrag aStillstandEintrag)
        {
            bool found = false;
            for (int i = 0; i < base.Count; i++)
            {
                if (base[i].Nr == aStillstandEintrag.Nr)
                {
                    found = true;
                    base[i] = aStillstandEintrag;
                    break;
                }
            }
            
            if (!found)
            {
                base.Add(aStillstandEintrag);
            }
            
            return base.Count - 1;
        }

        public int AddRaw(TStillstandEintrag aStillstandEintrag)
        {
            base.Add(aStillstandEintrag);
            return base.Count - 1;
        }

        public string getMaschNrsString()
        {
            System.Text.StringBuilder sb = new System.Text.StringBuilder();
            for (int i = 0; i < base.Count; i++)
            {
                if (i > 0) sb.Append(",");
                sb.Append(base[i].Maschnr);
            }
            return sb.ToString();
        }

        public TStillstandEintragsListe GetByMaschNr(int aMaschNr)
        {
            TStillstandEintragsListe result = new TStillstandEintragsListe();
            for (int i = 0; i < base.Count; i++)
            {
                if (base[i].Maschnr == aMaschNr)
                {
                    result.Add(base[i]);
                }
            }
            return result;
        }

        public int GetDauerByMaschNr(int aMaschNr)
        {
            int dauer = 0;
            for (int i = 0; i < base.Count; i++)
            {
                if (base[i].Maschnr == aMaschNr)
                {
                    TimeSpan span = base[i].Geht - base[i].Kommt;
                    dauer += (int)span.TotalMinutes;
                }
            }
            return dauer;
        }

        public int GetDauerByMaschNrFromDate(DateTime aFromDate, int aMaschNr)
        {
            int dauer = 0;
            for (int i = 0; i < base.Count; i++)
            {
                if (base[i].Maschnr == aMaschNr && base[i].Kommt >= aFromDate)
                {
                    TimeSpan span = base[i].Geht - base[i].Kommt;
                    dauer += (int)span.TotalMinutes;
                }
            }
            return dauer;
        }

        public TStillstandEintrag GetOpenByMaschNr(int aMaschNr)
        {
            for (int i = 0; i < base.Count; i++)
            {
                if (base[i].Maschnr == aMaschNr && base[i].Geht == DateTime.MinValue)
                {
                    return base[i];
                }
            }
            return null;
        }

        public new void Clear()
        {
            base.Clear();
        }
    }

    public class TStartStopEintrag
    {
        public string AuftragNr { get; set; } = "";
        public DateTime RuestStart { get; set; } = DateTime.MinValue;
        public DateTime Start { get; set; } = DateTime.MinValue;
        public DateTime Stop { get; set; } = DateTime.MinValue;

        private TStartStopEintrag CopyMe()
        {
            return new TStartStopEintrag
            {
                AuftragNr = this.AuftragNr,
                RuestStart = this.RuestStart,
                Start = this.Start,
                Stop = this.Stop
            };
        }
    }

    public class TStartStopEintragsListe : List<TStartStopEintrag>
    {
        public new TStartStopEintrag this[int index]
        {
            get => base[index];
            set => base[index] = value;
        }

        public int Add(TStartStopEintrag aStartStopEintrag)
        {
            base.Add(aStartStopEintrag);
            return base.Count - 1;
        }

        public TStartStopEintragsListe GetByBetriebsauftragNr(string aBANr)
        {
            TStartStopEintragsListe result = new TStartStopEintragsListe();
            for (int i = 0; i < base.Count; i++)
            {
                if (base[i].AuftragNr == aBANr)
                {
                    result.Add(base[i]);
                }
            }
            return result;
        }

        public new void Clear()
        {
            base.Clear();
        }
    }

    public class TSignalLogEintrag
    {
        public int maschnr { get; set; } = 0;
        public int wert { get; set; } = 0;
        public DateTime Start { get; set; } = DateTime.MinValue;
        public DateTime Stop { get; set; } = DateTime.MinValue;

        public TSignalLogEintrag CopyMe()
        {
            return new TSignalLogEintrag
            {
                maschnr = this.maschnr,
                wert = this.wert,
                Start = this.Start,
                Stop = this.Stop
            };
        }
    }

    public class TSignalLogEintragListe : List<TSignalLogEintrag>
    {
        public new TSignalLogEintrag this[int index]
        {
            get => base[index];
            set => base[index] = value;
        }

        public int Add(TSignalLogEintrag aSignalLogEintrag)
        {
            base.Add(aSignalLogEintrag);
            return base.Count - 1;
        }

        public TSignalLogEintragListe GetByMaschNr(int aMaschNr)
        {
            TSignalLogEintragListe result = new TSignalLogEintragListe();
            for (int i = 0; i < base.Count; i++)
            {
                if (base[i].maschnr == aMaschNr)
                {
                    result.Add(base[i]);
                }
            }
            return result;
        }

        public new void Clear()
        {
            base.Clear();
        }
    }
}
